compliant contact model supports autodiff

RussTedrake opened the pull request:
rigid_body_tree collision checks support autodiff, but throw if the gradient information is insufficient

this is the interesting one, which handles gradient information loss into/out of the collision engine.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8175)
<!-- Reviewable:end -->

RussTedrake commented:
+@seancurtis-tri for feature review.  (please pass to @jwnimmer-tri for platform if/when our discussion reaches consensus and I'm not online today).

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6H0TvNZO27w-J_pzGh:bi26rkt)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Note that I expect this to be RED on the build servers until #8171 lands, but preferred to not include that commit here.

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6H30J1_cLOtRhoeWoP:b-gdqw73)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
It isn't clear to me what the definition of "sufficient" is. What I infer is that you can use AutoDiff'd collision checks as long as things don't collide. Is that correct? I find the documentation is much vaguer about that with a repeated use of the pattern "may have been discarded". As implemented, I interpret this PR as "if scalar is AutoDiff and there are collisions, an exception gets thrown."

Unless there's an immediate follow-up PR that modifies the implementation into one more inline with what the documentation suggests, then I'd recommend bringing the documentation inline with the behavior. (Most of my comments below are on this same theme.)

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[multibody/rigid_body_tree.cc, line 1013 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HcMa27gJ27TdyO3xm:-L6HcMa27gJ27TdyO3xn:b-wda1l4) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.cc#L1013)):*
> ```cpp
>     auto& pair = contact_points[i];
>     if (pair.elementA->CanCollideWith(pair.elementB)) {
>       // For the autodiff version, throw if gradient information may have
> ```

BTW This comment isn't quite representative. The test assumes that if `U` is autodiff, then gradients *have* been lost (not "may have"). In other words, the existence of collisions seems to be sufficient to trigger the exception.

---

*[multibody/rigid_body_tree.h, line 1129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Hdf39lQ-8lzf8-h_P:-L6Hdf39lQ-8lzf8-h_Q:b-enrts4) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1129)):*
> ```objc
>    margins. If `false`, the representation without margins is used instead.
> 
>    @param throw_if_missing_gradient[in] If `true`, then the model will
> ```

BTW "[in]" should follow `@param`.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6HdmiEgHW1pjf37yVE:bmo5zcl) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
> ```objc
>    @param throw_if_missing_gradient[in] If `true`, then the model will
>    throw a std::runtime_error if the scalar type U is non-double and
>    gradient information may have been lost.
> ```

BTW To beat a dead horse: "may have been lost" makes it sound like there's a smart test to determine loss. What is the motivation for the distance between documentation and implementation?

---

*[multibody/rigid_body_plant/compliant_contact_model.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Hf8Lw74FYVaOIUQc6:-L6Hf8Lw74FYVaOIUQc7:b-6soll6) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.cc#L62)):*
> ```cpp
>   for (const auto& pair : pairs) {
>     if (pair.distance < 0.0) {  // There is contact.
>       if (!std::is_same<T, double>::value) {
> ```

FYI I find it curious that `false` was passed into the tree's `ComputeMaximumDepthCollisionPoints` method for the "throw on gradient lost" parameter. That leads to a duplication of code here. It *feels* like a redundant duplication. Looking at [bullet_model.cc](https://github.com/RobotLocomotion/drake/blob/master/multibody/collision/bullet_model.cc#L865) contacts would only be reported if the distance is less than zero. (Admittedly, the documentation of the method does *not* make this clear.)

So, why not simply pass `true` into the initial invocation?

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6HgQDKYKjuUr-sLm4y:b-cjzaoj) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
> ```objc
> /// CompliantContactModels use RigidBodyTree<double>.  This effectively implies
> /// that we can e.g. AutoDiffXd with respect to the configurations, but not
> /// the RigidBodyTree parameters.
> ```

FYI I find this comment a bit opaque. As I read it, I can autodiff w.r.t. geometry poses (clearly a function of configuration), but not geometry parameters (e.g., sphere radius). And yet this is *not* the case. I know this isn't what the intent of this comment is.

---

*[multibody/rigid_body_plant/test/compliant_contact_model_test.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HhICvlNwmfGgPKI6b:-L6HhICvlNwmfGgPKI6c:bgr6720) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/test/compliant_contact_model_test.cc#L154)):*
> ```cpp
> }
> 
> // Test that the autodiff module throws when collision information is discarded
> ```

FYI *This* comment seems to be the one that accurately represents the implemented behavior.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6HdGF2MV1bLSb9h-pC:bqx4fo0)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
tried to expand the comments... ptal.

---

Review status: 5 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[multibody/rigid_body_tree.cc, line 1013 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HcMa27gJ27TdyO3xm:-L6IIKJap8ByG-NA7qdS:b-ad5t4w) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.cc#L1013)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW This comment isn't quite representative. The test assumes that if `U` is autodiff, then gradients *have* been lost (not "may have"). In other words, the existence of collisions seems to be sufficient to trigger the exception.
</blockquote></details>

extended the comment.  there are cases where no information would have been lost.  (the simplest example is that this method does not distinguish between contact distance >= 0 or not)

---

*[multibody/rigid_body_tree.h, line 1129 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Hdf39lQ-8lzf8-h_P:-L6IM9kQHC8zOPP1hb4d:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1129)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW "[in]" should follow `@param`.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6IMFxYFjaoDndjyfJw:b-ms03s0) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW To beat a dead horse: "may have been lost" makes it sound like there's a smart test to determine loss. What is the motivation for the distance between documentation and implementation?
</blockquote></details>

because it really is possible (common) for many queries to have no relevant gradient information going into and out of the collision engine.  I hope my docs throughout are clearing that up.

---

*[multibody/rigid_body_plant/compliant_contact_model.cc, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Hf8Lw74FYVaOIUQc6:-L6IMbdLQHWjHlhezCyE:b-gm1w0e) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.cc#L62)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I find it curious that `false` was passed into the tree's `ComputeMaximumDepthCollisionPoints` method for the "throw on gradient lost" parameter. That leads to a duplication of code here. It *feels* like a redundant duplication. Looking at [bullet_model.cc](https://github.com/RobotLocomotion/drake/blob/master/multibody/collision/bullet_model.cc#L865) contacts would only be reported if the distance is less than zero. (Admittedly, the documentation of the method does *not* make this clear.)

So, why not simply pass `true` into the initial invocation?
</blockquote></details>

If that is true, then that would change things.  

But this works even if it is not true.  And I've added a comment to show another important case that I haven't handled yet, but could imagine handling within 2 weeks.

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6INhK3BxEjIaxdpKxt:b-4ope6c) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I find this comment a bit opaque. As I read it, I can autodiff w.r.t. geometry poses (clearly a function of configuration), but not geometry parameters (e.g., sphere radius). And yet this is *not* the case. I know this isn't what the intent of this comment is.
</blockquote></details>

In fact, that is the case.  

---

*[multibody/rigid_body_plant/test/compliant_contact_model_test.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HhICvlNwmfGgPKI6b:-L6IOINZvx7U_ypA7Bk6:b-88zori) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/test/compliant_contact_model_test.cc#L154)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI *This* comment seems to be the one that accurately represents the implemented behavior.
</blockquote></details>

glad you approve.  ;-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6I_uhTMXGLNNxKb4PM:bg7d7pq)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
And back in your court.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[multibody/rigid_body_tree.cc, line 1013 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HcMa27gJ27TdyO3xm:-L6J3G6fz3h3UTVsPf7H:bqaotpc) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.cc#L1013)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

extended the comment.  there are cases where no information would have been lost.  (the simplest example is that this method does not distinguish between contact distance >= 0 or not)
</blockquote></details>

My comment was unclear. This was just another iteration of my rejection of the word "may" as meaningful documentation. See other note.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6ItueNFRwmoVK2UgHg:b-nz0ce6) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

because it really is possible (common) for many queries to have no relevant gradient information going into and out of the collision engine.  I hope my docs throughout are clearing that up.
</blockquote></details>

I see what you mean. I still find the comments dissatisfying. It comes down to thinking about the semantics of the inputs. 

There is an implementation fact: no matter what, the returned autodiff values have gradient values that have, in some sense, "been stomped on" and rendered meaningless. That's a guarantee.

Your point is that if the input had garbage gradients, than transformation from one set of garbage to another is no loss.

At the end of the day, the user should know to pass `true` if they require meaningful gradients. Thus the context of invocation will define the value of the flag.

I don't think the documentation makes those semantics clear. Perhaps a change in parameter name: rather than describing gradients as being "lost" (which depends on the semantics of the caller), it should simply be `throw_if_bad_gradients` and then document that all gradients will be bad, and it is the caller's responsibility to decide if they care (throw an exception) or not (let it ride).

---

*[multibody/rigid_body_tree.h, line 937 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6It91umAsxAecarzj8:-L6It91vcS4kEXaNJ9IH:ba1v9zy) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L937)):*
BTW The boolean predicate seems misstated. What this says,

> if U != double && throw_if_missing_gradient --> casting && error

It seems to me that it *should* be:

> If U != double --> casting
> if casting && throw_if_missing_gradient --> error

---

*[multibody/rigid_body_tree.h, line 940 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6It5Y2uZMv3GzwwLep:-L6It5Y2uZMv3GzwwLeq:boienul) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L940)):*
> ```objc
>    * is `false`, then the transforms from kinematics cache will be forcefully 
>    * cast to doubles (discarding e.g. any gradient information).  Callers that
>    * set this option to false are responsible for ensuring  that all gradients 
> ```

BTW extraneous space between "ensuring" and "that".

---

*[multibody/rigid_body_tree.h, line 953 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Itaiy305cFnw-znIF:-L6Itaiy305cFnw-znIG:b-qxc8ih) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L953)):*
> ```objc
> 
>   /**
>    * @see updateCollisionElements 
> ```

BTW Does this render in doxygen? I'm always unsure about doxygen handling overloading.

---

*[multibody/rigid_body_tree.h, line 960 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Itl1iB39TnVcFbexR:-L6Itl1iB39TnVcFbexS:b-8ilzgw) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L960)):*
> ```objc
>   void updateDynamicCollisionElements(const KinematicsCache<U>& kin_cache,
>                                       bool throw_if_missing_gradient
>                                         = true);
> ```

FYI I'm quite surprised the `= true` doesn't fit on the same line.

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6IwSnACB-bgzb-f09b:b-jsr3dw) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

In fact, that is the case.  
</blockquote></details>

When I said, "this is not the case", I was suggesting that you can't autodiff w.r.t. *anything* through collision. No parameters and no configuration space. Or am I missing something?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6J4yVoR_lUcBqZnE9R:bdf6py8)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6JA5Ez-GCHt2i4BEUa:b-99x33e) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I see what you mean. I still find the comments dissatisfying. It comes down to thinking about the semantics of the inputs. 

There is an implementation fact: no matter what, the returned autodiff values have gradient values that have, in some sense, "been stomped on" and rendered meaningless. That's a guarantee.

Your point is that if the input had garbage gradients, than transformation from one set of garbage to another is no loss.

At the end of the day, the user should know to pass `true` if they require meaningful gradients. Thus the context of invocation will define the value of the flag.

I don't think the documentation makes those semantics clear. Perhaps a change in parameter name: rather than describing gradients as being "lost" (which depends on the semantics of the caller), it should simply be `throw_if_bad_gradients` and then document that all gradients will be bad, and it is the caller's responsibility to decide if they care (throw an exception) or not (let it ride).
</blockquote></details>

i still disagree with your view.  in some cases, the user does not require the component of the gradients that the collision detection unit provides -- because they would clearly be zero or otherwise have no effect.  the only missing gradient is the point on the body (in body coordinates) relative to the configuration.  So if, hypothetically we were to have two points collide (I understand this is a measure zero event, but work with me), then we know that the gradient must be zero even if the collision engine didn't tell us so.  So the gradients are actually correct.  In fact, they will be correct any time that the location of the collision point on the body does not change with the configuration.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6JBejqUrZ0pKoC9Q09:b-trlwwr) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i still disagree with your view.  in some cases, the user does not require the component of the gradients that the collision detection unit provides -- because they would clearly be zero or otherwise have no effect.  the only missing gradient is the point on the body (in body coordinates) relative to the configuration.  So if, hypothetically we were to have two points collide (I understand this is a measure zero event, but work with me), then we know that the gradient must be zero even if the collision engine didn't tell us so.  So the gradients are actually correct.  In fact, they will be correct any time that the location of the collision point on the body does not change with the configuration.
</blockquote></details>

I'll try to rephrase it with something more technical like "zero gradients" instead of "bad" or "missing" when I'm back at my dev machine in a little bit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6JD2pUnVL-n2VmO485:b1aj8p1) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I'll try to rephrase it with something more technical like "zero gradients" instead of "bad" or "missing" when I'm back at my dev machine in a little bit.
</blockquote></details>

The explanation makes sense. It reveals one of the assumptions we apparently didn't have in common: that the double -> AutoDiff conversion implicitly initializes derivatives to be zero.

Is it guaranteed to be zero?  For both variable size derivative vectors as well as fixed-size? This should appear in the unit test. (And be documented as you allude to.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6JFqpvHaeaPmPkaLrI:bfkmacv) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

The explanation makes sense. It reveals one of the assumptions we apparently didn't have in common: that the double -> AutoDiff conversion implicitly initializes derivatives to be zero.

Is it guaranteed to be zero?  For both variable size derivative vectors as well as fixed-size? This should appear in the unit test. (And be documented as you allude to.)
</blockquote></details>

It’s slightly more subtle.  I am casting from double to autodiff.  We do this absolutely everywhere in the code... sometimes explicitly, sometimes implicitly.  In this specific case, (anywhere with autodiffxd,) the cast variable has a derivative vector of length zero.  Performing operations on this with other sutodiff results in it being treated effectively as a constant.

This is autodiff behavior... which we should cover with tests (and may already?) but probably not here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6JNU6G9wlO_xIzWs5V:bnu3vnq) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

It’s slightly more subtle.  I am casting from double to autodiff.  We do this absolutely everywhere in the code... sometimes explicitly, sometimes implicitly.  In this specific case, (anywhere with autodiffxd,) the cast variable has a derivative vector of length zero.  Performing operations on this with other sutodiff results in it being treated effectively as a constant.

This is autodiff behavior... which we should cover with tests (and may already?) but probably not here?
</blockquote></details>

I knew that about the the `AutoDiffScalar` instantiated on a `VectorX`. I'm not so sure about the fixed-size vectors. So, in the case of a size-zero `VectorX` it is *functionally* all zeroes. Is it the same for fixed-size?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Back to you.  
Note: contains #8202.
also resolved the remaining linker errors since we decided not to zap the old RBtree code from the other PR.

---

Review status: 6 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[multibody/rigid_body_tree.cc, line 1013 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HcMa27gJ27TdyO3xm:-L6QoOPYqRLa_OCddLdM:bdzi4sn) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.cc#L1013)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

My comment was unclear. This was just another iteration of my rejection of the word "may" as meaningful documentation. See other note.
</blockquote></details>

Done. (I hope)

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6Qm15XRu8_3Khk5kug:b-qy9261) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I knew that about the the `AutoDiffScalar` instantiated on a `VectorX`. I'm not so sure about the fixed-size vectors. So, in the case of a size-zero `VectorX` it is *functionally* all zeroes. Is it the same for fixed-size?
</blockquote></details>

I've added a test to convince you.  I hadn't actually checked (better than I did), but in my mind that is a bit like suggesting that casting from int to double could leave garbage values after the decimal place.  I would be really sad if that was the case.  ;-)

see #8202.

---

*[multibody/rigid_body_tree.h, line 937 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6It91umAsxAecarzj8:-L6R5SrYLVt-DUGMggLP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L937)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW The boolean predicate seems misstated. What this says,

> if U != double && throw_if_missing_gradient --> casting && error

It seems to me that it *should* be:

> If U != double --> casting
> if casting && throw_if_missing_gradient --> error
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 940 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6It5Y2uZMv3GzwwLep:-L6R64H2kdEKkBcJBH_w:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L940)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW extraneous space between "ensuring" and "that".
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 953 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Itaiy305cFnw-znIF:-L6R90WG3svy2kDxnCOb:b-wpd7m6) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L953)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Does this render in doxygen? I'm always unsure about doxygen handling overloading.
</blockquote></details>

yup.  renders fine.

---

*[multibody/rigid_body_tree.h, line 960 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Itl1iB39TnVcFbexR:-L6R6FXg9_iSiBiYPmLd:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/99be6988b27e04119ad5635beaa40d13477169ce/multibody/rigid_body_tree.h#L960)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I'm quite surprised the `= true` doesn't fit on the same line.
</blockquote></details>

Done.

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6QmcJ9z4J8uKJktIUq:b3govug) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

When I said, "this is not the case", I was suggesting that you can't autodiff w.r.t. *anything* through collision. No parameters and no configuration space. Or am I missing something?
</blockquote></details>

You can autodiff through the contact model w/ respect to configurations of the tree, with the exception of the collision engine (collision != contact model).  i added this line in response to your earlier questions about why it was ok to have RigidBodyTree<double>.  honestly, not sure how to write it better.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6RAEePy7AA1o1byPu2:bnufzdr)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Some cosmetic stuff and a bit more discussion. Still have CI errors.

+@soonho-tri for platform review for tomorrow.

---

Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[math/autodiff_gradient.h, line 146 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Rtrp9TDAL3zNlBkp3:-L6Rtrp9TDAL3zNlBkp4:b99je5q) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/autodiff_gradient.h#L146)):*
> ```objc
>  *   `B = A.cast<double>()`
>  * should (and does) fail to compile.  Use `CastToValueOrThrow(A)` if you want
>  * to
> ```

FYI stray newline

---

*[math/test/autodiff_test.cc, line 127 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6S0zt857Lh72NUoNMb:-L6S0zt857Lh72NUoNMc:b-5g6svq) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/test/autodiff_test.cc#L127)):*
> ```cpp
> }
> 
> GTEST_TEST(AdditionalAutodiffTest, CastToValueOrThrow) {
> ```

BTW You have several calls to `CastToValueOrThrow` which you expect to *not* throw an exception. It would probably be good to wrap those in an `EXPECT_NO_THROW` even though it plays around with RAII. It does formalize the expectation.

---

*[math/test/autodiff_test.cc, line 133 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6RudoOVxiBGUIFsBWl:-L6RudoOVxiBGUIFsBWm:b-h60mt6) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/test/autodiff_test.cc#L133)):*
> ```cpp
> 
>   Eigen::MatrixXd test2 = Eigen::Vector3d{1., 2., 3.};
>   EXPECT_TRUE(CompareMatrices(CastToValueOrThrow(test2), test2));
> ```

FYI Given the declared interface for doubles returns a reference, we could test that by comparing addresses of inputs and outputs? Adding the test would catch changes to that API.

---

*[math/test/autodiff_test.cc, line 189 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6S2O0NPQua-lnqqnDP:-L6S2O0NPQua-lnqqnDQ:b-siflcw) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/test/autodiff_test.cc#L189)):*
> ```cpp
>   EXPECT_EQ(fixed_gradients.rows(), 2);
>   EXPECT_EQ(fixed_gradients.cols(), 3);
>   EXPECT_TRUE(fixed_gradients.isZero(1e-8));
> ```

BTW I would've expected this test would pass with a precision of zero.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6S4ppj5oKmtlxgotR6:b-rtdu7i) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I've added a test to convince you.  I hadn't actually checked (better than I did), but in my mind that is a bit like suggesting that casting from int to double could leave garbage values after the decimal place.  I would be really sad if that was the case.  ;-)

see #8202.
</blockquote></details>

Touche.  :)

---

*[multibody/rigid_body_tree.h, line 1141 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6S4rs-ewwx7SNtZlh1:-L6S4rs-ewwx7SNtZlh2:b9r4u46) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/multibody/rigid_body_tree.h#L1141)):*
This statement doesn't quite parse right. Perhaps:

> @throws std::runtime_error based on the criteria of CastToValueOrThrow(), but only if @p throws_if_missing_gradient is `true`.

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6S5HS1M6HbIB95oapB:b-lm3t37) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

You can autodiff through the contact model w/ respect to configurations of the tree, with the exception of the collision engine (collision != contact model).  i added this line in response to your earlier questions about why it was ok to have RigidBodyTree<double>.  honestly, not sure how to write it better.  
</blockquote></details>

I'm not going to hold this PR up on this point. I guess I still don't quite see the model in your mind. So, I'm going to propagate this conversation on the side for my own education.

In my mind, because the collision engine is the pinch in the contact model's hourglass, all gradients go to hell (more or less). Whatever beautiful gradients may come in from the `KinematicsCache` are wiped out when computing contact forces -- the forces can't report in any way shape or form how they depend on the configuration variables. And that loss propagates through. 

Now, all of this, of course, is dependent on the gradients. If the gradient is defined by the partial w.r.t. *velocities*, then we don't lose any information and all of the zero gradients coming back out of collision are fine (because none of the collision queries depend on velocities). But if the gradient is partials w.r.t. *position*, it's all gone to pot.

Is that right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6SA6sv7gCkXsDKBkW7:b-c6djg)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
> please pass to @jwnimmer-tri for platform if/when our discussion reaches consensus and I'm not online today

@jwnimmer-tri , do you want to handle this as requested?

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6SNju5zbquAtuwJF1A:bx232d6)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I believe that was predicated on the original timing (i.e., Monday). But I'll let you guys work it out.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6STv-CZBrqGmQdX0HP:b-apif7i)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Got it. I'll handle it this afternoon.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6SVjQuUbhzcoRRp6nY:b-asee1h)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
With the CI failure, probably alright to wait until your regular day tomorrow.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6SWDGDRIkIqCIeJXJ5:b8yiz9c)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
@soonho-tri getting your feedback today would be awesome.  i'll fix the build issues and push into the nightly tonight if i can muster it.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6SjYgKwOGnC76FuJJe:bfx1g4s)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@russtedrake , copy that.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6Sk8rh7Gn6vCGq_MWz:bj5llo2)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[math/autodiff_gradient.h, line 142 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6TCAjtcjnwwTyKKaS5:-L6TQYliIln5I1b4cjP3:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/autodiff_gradient.h#L142)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, I think it's better to mention the role of `precision` argument here explicitly.
</blockquote></details>

Done.

---

*[math/autodiff_gradient.h, line 146 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6Rtrp9TDAL3zNlBkp3:-L6TQY-2dVstl9i9HpeK:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/autodiff_gradient.h#L146)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI stray newline
</blockquote></details>

Done.

---

*[math/autodiff_gradient.h, line 182 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6TE9sV8wRv7wcqgxXw:-L6TP_cZb5wXwzei2FZA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/autodiff_gradient.h#L182)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, `g++-5` complains about the unused variable `precision`. `double = 0.` should resolve this issue. (Ditto below)
</blockquote></details>

Done.

---

*[math/test/autodiff_test.cc, line 127 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6S0zt857Lh72NUoNMb:-L6TRfG3G5o5Z8-bJfB6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/test/autodiff_test.cc#L127)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW You have several calls to `CastToValueOrThrow` which you expect to *not* throw an exception. It would probably be good to wrap those in an `EXPECT_NO_THROW` even though it plays around with RAII. It does formalize the expectation.
</blockquote></details>

Done.

---

*[math/test/autodiff_test.cc, line 133 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6RudoOVxiBGUIFsBWl:-L6TSAvLcUsnYR9-c9gb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/test/autodiff_test.cc#L133)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI Given the declared interface for doubles returns a reference, we could test that by comparing addresses of inputs and outputs? Adding the test would catch changes to that API.
</blockquote></details>

Done.

---

*[math/test/autodiff_test.cc, line 189 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6S2O0NPQua-lnqqnDP:-L6TSf4zABHCVKhc2g2p:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/math/test/autodiff_test.cc#L189)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I would've expected this test would pass with a precision of zero.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 1131 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HdmiEgHW1pjf37yVD:-L6TTN1QNPmJ0bLQrcls:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_tree.h#L1131)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Touche.  :)
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 1141 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6S4rs-ewwx7SNtZlh1:-L6TTLCq0c4Jqu-4ooJY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/739f7aea8dd0d6ce25c238c5bfd7c0460fdb56da/multibody/rigid_body_tree.h#L1141)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I'd refine @seancurtis-tri 's suggestion further into:

> @throws std::runtime_error based on the criteria of CastToValueOrThrow() only if @p throws_if_missing_gradient is true.


</blockquote></details>

Done.

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6TTVZDJGRuUNAEy5mY:b-xhq6ed) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'm not going to hold this PR up on this point. I guess I still don't quite see the model in your mind. So, I'm going to propagate this conversation on the side for my own education.

In my mind, because the collision engine is the pinch in the contact model's hourglass, all gradients go to hell (more or less). Whatever beautiful gradients may come in from the `KinematicsCache` are wiped out when computing contact forces -- the forces can't report in any way shape or form how they depend on the configuration variables. And that loss propagates through. 

Now, all of this, of course, is dependent on the gradients. If the gradient is defined by the partial w.r.t. *velocities*, then we don't lose any information and all of the zero gradients coming back out of collision are fine (because none of the collision queries depend on velocities). But if the gradient is partials w.r.t. *position*, it's all gone to pot.

Is that right?
</blockquote></details>

I don't mind the discussion (but admit I will be happy to separate it from this review).  It's not an hour-glass.  The gradients of the collision element pose w/ respect to the autodiff variables stay intact.  It is combined (via the chain rule) with a the location of the contact point relative to the body, which here is set arbitrarily to zero -- which then acts like a constant in the chain rule computations.  The rest of the gradients are intact.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6ThebZulneqTnnl0Mg:botryjw) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

I don't mind the discussion (but admit I will be happy to separate it from this review).  It's not an hour-glass.  The gradients of the collision element pose w/ respect to the autodiff variables stay intact.  It is combined (via the chain rule) with a the location of the contact point relative to the body, which here is set arbitrarily to zero -- which then acts like a constant in the chain rule computations.  The rest of the gradients are intact.  
</blockquote></details>

Thanks for the elaboration and humoring me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/rigid_body_plant/compliant_contact_model.h, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-L6HgQDKYKjuUr-sLm4x:-L6Ttnyjgz1ggLT-aK7G:b-k0nioc) ([raw file](https://github.com/robotlocomotion/drake/blob/68ae3ef745b7c9814ed31d809b2eaa5cdfe6ae78/multibody/rigid_body_plant/compliant_contact_model.h#L33)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Thanks for the elaboration and humoring me.
</blockquote></details>

shortest version is that we have 
y = transform*pt.    
dydx = dtransformdx*pt + transform*dptdx.  
the dptdx is zero because of the collision detector, but the first term is fine (and in this case, pt is just the pt on the body, so for e.g. point collision geometry, it's zero anyways).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

FTR, my only concern is that in `math/{autodiff,autodiff_gradient}.h` we have this binary assumption; "`ScalarType` is either `double` or `AutoDiff`" which does not hold considering `symbolic::Expression`. For now, a compiler will stop users from instantiating `DiscardGradient`/`DiscardZeroGradient` with a matrix of `symbolic::Expression`, so I consider it a lack of feature, not a defect.  I'll see if I can fix it later this week. Russ pointed out this to me this morning as well.

---

Reviewed 4 of 9 files at r1, 1 of 7 files at r3, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6U0uF90ieM3sgzQPCO:b-nwpf2z)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
exactly.  and in general, it would be good to have an "isa autodiffscalar<anything>" check...  Sean and I chatted about it last week and agreed that it wasn't simple.  thanks both.

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/8175#-:-L6U4ozyJOl1qzJwqN0g:b4xb13x)*
<!-- Sent from Reviewable.io -->
